tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/Qualification.scala

```diff
@@ -28,31 +31,71 @@ import org.apache.spark.sql.rapids.tool.qualification._
  * Scores the applications for GPU acceleration and outputs the
  * reports.
  */
-object Qualification extends Logging {
+class Qualification(outputDir: String, numRows: Int, hadoopConf: Configuration,
+    timeout: Option[Long], nThreads: Int) extends Logging {
 
-  def qualifyApps(
-      allPaths: Seq[EventLogInfo],
-      numRows: Int,
-      outputDir: String,
-      hadoopConf: Configuration): ArrayBuffer[QualificationSummaryInfo] = {
-    val allAppsSum: ArrayBuffer[QualificationSummaryInfo] = ArrayBuffer[QualificationSummaryInfo]()
+  private val allApps = new ConcurrentLinkedQueue[QualificationSummaryInfo]()
+  // default is 24 hours
+  private val waitTimeInSec = timeout.getOrElse(60 * 60 * 24L)
+
+  private val threadFactory = new ThreadFactoryBuilder()
+    .setDaemon(true).setNameFormat("qualTool" + "-%d").build()
+  logInfo(s"Threadpool size is $nThreads")
+  private val threadPool = Executors.newFixedThreadPool(nThreads, threadFactory)
+    .asInstanceOf[ThreadPoolExecutor]
+
+  private class QualifyThread(path: EventLogInfo) extends Runnable {
+    def run: Unit = qualifyApp(path, numRows, hadoopConf)
+  }
+
+  def qualifyApps(allPaths: Seq[EventLogInfo]): Seq[QualificationSummaryInfo] = {
     allPaths.foreach { path =>
+      try {
+        threadPool.submit(new QualifyThread(path))
+      } catch {
+        case e: Exception =>
+          logError(s"Unexpected exception submitting log ${path.eventLog.toString}, skipping!", e)
+      }
+    }
+    // wait for the threads to finish processing the files
+    threadPool.shutdown()
+    if (!threadPool.awaitTermination(waitTimeInSec, TimeUnit.SECONDS)) {
+      logError(s"Processing log files took longer then $waitTimeInSec seconds," +
+        " stopping processing any more event logs")
+      threadPool.shutdownNow()
+    }
```

Comment:
we may get an InterruptedException in awaitTermination, may want to surround it with try/finally and place shutdownNow into finally.